internal/bridge/client_test.go

```diff
@@ -0,0 +1,450 @@
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	gocmp "github.com/google/go-cmp/cmp"
+	gocmpopts "github.com/google/go-cmp/cmp/cmpopts"
+	"gotest.tools/v3/assert"
+)
+
+// TestClientBackoff logs the backoff timing chosen by [NewClient] for use
+// with `go test -v`.
+func TestClientBackoff(t *testing.T) {
+	client := NewClient("", "")
+	var total time.Duration
+
+	for i := 1; i <= 50 && client.Backoff.Steps > 0; i++ {
+		step := client.Backoff.Step()
+		total += step
+
+		t.Logf("%02d:%20v%20v", i, step, total)
+	}
+}
+
+func TestClientURL(t *testing.T) {
+	assert.Equal(t, defaultAPI, NewClient("", "").BaseURL.String())
+	assert.Equal(t, defaultAPI, NewClient("/path", "").BaseURL.String())
+	assert.Equal(t, defaultAPI, NewClient("http://:9999", "").BaseURL.String())
+	assert.Equal(t, defaultAPI, NewClient("postgres://localhost", "").BaseURL.String())
+	assert.Equal(t, defaultAPI, NewClient("postgres://localhost:5432", "").BaseURL.String())
+
+	assert.Equal(t,
+		"http://localhost:12345", NewClient("http://localhost:12345", "").BaseURL.String())
+}
+
+func TestClientDoWithBackoff(t *testing.T) {
+	t.Run("Arguments", func(t *testing.T) {
+		var bodies []string
+		var requests []http.Request
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			body, _ := io.ReadAll(r.Body)
+			bodies = append(bodies, string(body))
+			requests = append(requests, *r)
+
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write([]byte(`some-response`))
+		}))
+		t.Cleanup(server.Close)
+
+		// Client with one attempt, i.e. no backoff.
+		client := NewClient(server.URL, "xyz")
+		client.Backoff.Steps = 1
+		assert.Equal(t, client.BaseURL.String(), server.URL)
+
+		ctx := context.Background()
+		response, err := client.doWithBackoff(ctx,
+			"ANY", "/some/path", []byte(`the-body`),
+			http.Header{"Some": []string{"header"}})
+
+		assert.NilError(t, err)
+		assert.Assert(t, response != nil)
+		t.Cleanup(func() { _ = response.Body.Close() })
+
+		// Arguments became Request fields, including the client version.
+		assert.Equal(t, len(requests), 1)
+		assert.Equal(t, bodies[0], "the-body")
+		assert.Equal(t, requests[0].Method, "ANY")
+		assert.Equal(t, requests[0].URL.String(), "/some/path")
+		assert.DeepEqual(t, requests[0].Header.Values("Some"), []string{"header"})
+		assert.DeepEqual(t, requests[0].Header.Values("User-Agent"), []string{"PGO/xyz"})
+
+		body, _ := io.ReadAll(response.Body)
+		assert.Equal(t, string(body), "some-response")
+	})
+
+	t.Run("Idempotency", func(t *testing.T) {
+		var bodies []string
+		var requests []http.Request
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			body, _ := io.ReadAll(r.Body)
+			bodies = append(bodies, string(body))
+			requests = append(requests, *r)
+
+			switch len(requests) {
+			case 1, 2:
+				w.WriteHeader(http.StatusBadGateway)
+			default:
+				w.WriteHeader(http.StatusNotAcceptable)
```

Comment:
Correct. At most five will/should happen because of `client.Backoff.Steps = 5` below.

🤔 Should I put anything here to make that clear?

```suggestion
				w.WriteHeader(http.StatusBadGateway)
			case 3, 4, 5:
				w.WriteHeader(http.StatusNotAcceptable)
			default:
				t.Fatal("BUG: expected less than five requests")
```

Reply:
Given the code, anything more than 3 tries should be a fatal error, yeah?

Reply:
Yeah. The expectation is certainly three. Some thoughts below, but **tl;dr** I like the `1,2` with `Assert(3)` as-is.

---

I can't decide if/how to communicate the relationships between these places.

https://github.com/CrunchyData/postgres-operator/blob/3b9df4abb6c1642d92c5bdefe3a47cc55a8c9f20/internal/bridge/client_test.go#L116
https://github.com/CrunchyData/postgres-operator/blob/3b9df4abb6c1642d92c5bdefe3a47cc55a8c9f20/internal/bridge/client_test.go#L128

I typed out the following, but I don't like it:

```go
		const expected = 3
		…
		server := httptest.NewServer(… {
			…
			if len(requests) < expected {
				w.WriteHeader(http.StatusBadGateway)
			} else {
				w.WriteHeader(http.StatusNotAcceptable)
			}
		})
		…

		client := NewClient(server.URL, "")
		client.Backoff.Duration = time.Millisecond
		client.Backoff.Steps = 2 * expected
		…

		assert.Equal(t, len(requests), expected)
```

Reply:
I think it's clear as is -- if anything, I just had to refresh myself on http codes and look back to the behavior in `client.go` around which return codes retry and which exit.

Reply:
I struggled with http codes as well. I [mostly] used numbers and strings instead of named constants in client.go because the docs I was matching do as well.

The one place I didn't is below. I'm going to change it to a number now because the comparison is not strict equality.

https://github.com/CrunchyData/postgres-operator/blob/76d8a40ff9bfb687e67927c524d4238b94dc69a7/internal/bridge/client.go#L112-L113